### PR TITLE
Refactor dagger mains for better extensibility

### PR DESCRIPTION
**Pull request by devinrsmith**
This PR moves JettyMain from `server-jetty` to `server-jetty-app`; the main class isn't really meant to be a "library". That said, it's a very few short lines of code for integrators to achieve the same thing that JettyMain does:

```java
public final class MyCustomMain {
    public static void main(String[] args)
            throws IOException, InterruptedException, ClassNotFoundException, TimeoutException {
        final Configuration configuration = MainHelper.init(args, MyCustomMain.class);
        new CommunityComponentFactory()
                .build(configuration)
                .getServer()
                .run()
                .join();
    }
}
```

This PR also uses the "Community" prefix in more places to hint to integrators that it may or may not be appropriate for "Enterprise" contexts. 

This PR also introduces `server-jetty-app-custom` as an example for how an integrator can customize the structure and configuration of the Deephaven server. It's also useful from the perspective of dogfooding and ensuring DHC developers are aware of any changes that may break integrators.

As part of the above, `ComponentFactoryBase` has been added as an abstraction around creating Components. While not technically necessary, it's a useful construct that allows constructing components without callers needing to directly invoke dagger.

### CODE REVIEW

File: server/jetty-app-custom/src/main/java/io/deephaven/server/custom/CustomApplication1.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import io.deephaven.appmode.ApplicationState;
+import io.deephaven.appmode.ApplicationState.Listener;
+import io.deephaven.engine.liveness.LivenessScope;
+import io.deephaven.engine.liveness.LivenessScopeStack;
+import io.deephaven.engine.util.TableTools;
+import io.deephaven.util.SafeCloseable;
+
+import java.util.Objects;
+
+public final class CustomApplication1 implements ApplicationState.Factory {
```
Comment (nbauernfeind) [inline server/jetty-app-custom/src/main/java/io/deephaven/server/custom/CustomApplication1.java:15]: Please add javadoc that makes clear what to expect from this application.

File: server/jetty-app-custom/src/main/java/io/deephaven/server/custom/CustomApplication2.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+package io.deephaven.server.custom;
+
+import io.deephaven.appmode.ApplicationState;
+import io.deephaven.appmode.ApplicationState.Listener;
+import io.deephaven.engine.liveness.LivenessScope;
+import io.deephaven.engine.liveness.LivenessScopeStack;
+import io.deephaven.engine.util.TableTools;
+import io.deephaven.util.SafeCloseable;
+
+public final class CustomApplication2 implements ApplicationState.Factory {
```
Comment (nbauernfeind) [inline server/jetty-app-custom/src/main/java/io/deephaven/server/custom/CustomApplication2.java:13]: Please add javadoc that makes clear what to expect from this application.

File: server/jetty-app-custom/src/main/java/io/deephaven/server/custom/CustomAuthorization.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.server.custom;
+
+import io.deephaven.auth.codegen.impl.InputTableServiceContextualAuthWiring;
+import io.deephaven.server.auth.CommunityAuthorizationProvider;
+
+import javax.inject.Inject;
+
+public final class CustomAuthorization extends CommunityAuthorizationProvider {
```
Comment (nbauernfeind) [inline server/jetty-app-custom/src/main/java/io/deephaven/server/custom/CustomAuthorization.java:11]: Please add javadoc that makes clear what to expect from using this CustomAuthorization implementation.

File: server/src/main/java/io/deephaven/server/runner/MainHelper.java [hunk 4]
```diff
@@ -33,7 +32,7 @@
 import java.nio.file.Path;
 import java.util.Optional;
 
-public class Main {
+public class MainHelper {
```
Comment (nbauernfeind) [inline server/src/main/java/io/deephaven/server/runner/MainHelper.java:35]: Thanks for the rename here!

**Review by nbauernfeind — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/1997
Conceptual: https://github.com/deephaven/deephaven.io/issues/1998
Reference: https://github.com/deephaven/deephaven.io/issues/1996
Blog: https://github.com/deephaven/deephaven.io/issues/1995

**Pull request merged**